onnxruntime/core/graph/graph_proto_serializer.cc

```diff
@@ -0,0 +1,46 @@
+#include "core/graph/graph_proto_serializer.h"
+
+namespace onnxruntime {
+
+void GraphProtoSerializer::ToProto(ONNX_NAMESPACE::GraphProto& graph_proto, bool include_initializer) {
+  graph_proto.set_name(graph_viewer_->Name());
+  graph_proto.set_doc_string(graph_viewer_->Description());
+
+  for (const auto* input_arg : graph_viewer_->GetInputsIncludingInitializers()) {
+    *(graph_proto.mutable_input()->Add()) = input_arg->ToProto();
+  }
+
+  for (const auto* output_arg : graph_viewer_->GetOutputs()) {
+    *(graph_proto.mutable_output()->Add()) = output_arg->ToProto();
+  }
+
+  for (const auto* value_info : graph_viewer_->GetValueInfo()) {
+    *(graph_proto.mutable_value_info()->Add()) = value_info->ToProto();
+  }
+
+  // add the NodeArg info for outer scope NodeArgs so we capture the type information
+  for (const auto& name : graph_viewer_->GetOuterScopeNodeArgNames()) {
```

Comment:
Not sure what the correct treatment of these is - depends on how the GraphProto that is created will be used. 

If it is a standalone Graph they probably need to become inputs. 

If it's a subgraph you may not need to add value info entries as the types/shapes should be able to be inferred from the parent graph.

Reply:
yes i am not very clear what does the outer scope node arg means here. is it only used for the subgraphs in control ops?
my guess is if subgraph maps to a code block in the program. the "outer scope node args" is like some gloabl variables that the code block can refer to. 
But here, to serialize to protobuf, if we don't include it, will it make the graph proto invalid?

Reply:
so when perform the Compile on subgraph, if EP want to seriliaze the subgraph into graphproto, should we serialize the outer scope value into the graphproto? i am not quite sure about the behavior.

Reply:
Depends on what you want to do with the GraphProto. Something has to provide access to these values to execute the graph. If you want to treat it as a standalone Graph that can be called directly it would need any outer scope values to be converted into explicit graph inputs so that access. 

For example, say I'm dealing with the body of a Scan and I want my EP to compile that. Inside the subgraph of the Scan is an If node with 2 subgraphs. I would expect that we serialize the entire Scan body so that it becomes a standalone graph. Any outer scope inputs to the Scan body would need to become explicit inputs to this new standalone graph. If we serialize the If node as part of the Scan, I would expect it can still reference outer scope values given it's still contained within the Scan graph.